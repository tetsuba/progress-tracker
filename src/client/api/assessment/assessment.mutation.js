import gql from 'graphql-tag'

export const ADD_ASSESSMENT_MUTATION = gql`
  mutation AddAssessment($input: AddAssessmentInput!) {
    addAssessment(input: $input) {
      firstName
      assessments
    }
  }
`
