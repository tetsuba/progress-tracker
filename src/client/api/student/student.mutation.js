import gql from 'graphql-tag'

export const ADD_STUDENT_MUTATION = gql`
  mutation AddStudent($input: AddStudentInput!) {
    addStudent(input: $input) {
      _id
      firstName
      lastName
    }
  }
`
