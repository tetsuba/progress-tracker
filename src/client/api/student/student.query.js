import gql from 'graphql-tag'

export const STUDENTS_QUERY = gql`
  {
    students {
      _id
      firstName
      lastName
    }
  }
`

export const STUDENT_QUERY = gql`
  query($input: StudentIdInput!) {
    getStudent(input: $input) {
      firstName
      lastName
      assessments
    }
  }
`
