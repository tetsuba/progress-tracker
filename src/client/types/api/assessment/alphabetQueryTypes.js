type assessmentAlphabetHistoryAlphabetType = {
  value: number,
  letter: string,
}

type assessmentAlphabetHistoryType = {
  date: string,
  alphabet: [assessmentAlphabetHistoryAlphabetType],
}

export type AssessmentAlphabetType = {
  getStudentCourseABC: {
    studentName: string,
    history: [assessmentAlphabetHistoryType],
  },
}
