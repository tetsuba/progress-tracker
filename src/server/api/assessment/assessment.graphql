
input AddAssessmentInput {
  studentId: String
  assessmentName: String,
}

extend type Mutation {
  addAssessment(input: AddAssessmentInput!): StudentType!
}