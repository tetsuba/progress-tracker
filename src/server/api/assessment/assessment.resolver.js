const { errorName } = require('../../errorHandling')

module.exports = {
  Query: {},
  Mutation: {
    addAssessment: async (_, args, context, info) => {
      const { Student, AlphabetAssessment } = context.models
      const { studentId, assessmentName } = args.input

      const student = await Student.findById(studentId)
      if (!student) {
        throw new Error(errorName.STUDENT_NOT_FOUND)
      }

      student.assessments.push(assessmentName)
      await student.save()

      switch (assessmentName.split('-')[0]) {
        case 'alphabet':
          // TODO: how to test an assessment was created?

          console.log('assessmentName:', assessmentName)
          console.log('AlphabetAssessment:', AlphabetAssessment)

          await AlphabetAssessment.create({
            studentId: studentId,
            typeOfAlphabetAssessment: assessmentName.split('-')[1],
          })
          return student

        default:
          throw new Error(errorName.ASSESSMENT_NOT_FOUND)
      }
    },
  },
}
