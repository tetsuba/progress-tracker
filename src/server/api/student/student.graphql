
type StudentType {
  id: String
  firstName: String
  lastName: String
  DOB: String
  assessments: [String]
}

type StudentsType {
  _id: String
  firstName: String
  lastName: String
}

input StudentIdInput {
  id: String
}

input AddStudentInput {
  firstName: String
  lastName: String
  DOB: String
}

extend type Query {
  students: [StudentsType]!
  getStudent(input: StudentIdInput!): StudentType!
}

extend type Mutation {
  addStudent(input: AddStudentInput!): [StudentsType]!
}
