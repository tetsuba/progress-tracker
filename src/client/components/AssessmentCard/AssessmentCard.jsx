import React from 'react'
import { useParams } from 'react-router-dom'
import { useMutation } from '@apollo/react-hooks'
import { Button, Col } from 'react-bootstrap'

// MUTATIONS
import { ADD_ASSESSMENT_MUTATION } from '../../api/assessment/assessment.mutation'

// QUERIES
import { STUDENT_QUERY } from '../../api/student/student.query'

// ICONS
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlus } from '@fortawesome/free-solid-svg-icons'

// TYPES
import type { AssessmentType } from '../AssessmentList/AssessmentList.types'

type Props = {
  assessment: AssessmentType,
}

export default function AssessmentCard(props: Props) {
  const { title, description, assessmentId } = props.assessment
  const { id }: { id?: ?string } = useParams()
  const variables = {
    input: {
      id: id,
    },
  }
  const [addAssessment] = useMutation(ADD_ASSESSMENT_MUTATION, {
    refetchQueries: [{ query: STUDENT_QUERY, variables }],
  })

  return (
    <Col sm={4} key={title}>
      <div className="card text-body text-decoration-none">
        <div className="card-header">
          <h5>{title}</h5>
        </div>
        <div className="card-body">{description}</div>
        <div className="card-footer">
          <Button
            className="float-right"
            variant="primary"
            onClick={() => {
              const options = {
                variables: {
                  input: {
                    studentId: id,
                    assessmentName: assessmentId,
                  },
                },
              }
              addAssessment(options).catch(() => console.log('error!'))
            }}
          >
            <FontAwesomeIcon icon={faPlus} /> Add Assessment
          </Button>
        </div>
      </div>
    </Col>
  )
}
