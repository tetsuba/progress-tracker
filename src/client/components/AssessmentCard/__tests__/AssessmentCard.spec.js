import { act } from 'react-dom/test-utils'
import { delay, graphRenderer } from '../../../../test/testHelper'
import AssessmentCard from '../AssessmentCard'
import { addAssessmentMutationSuccess } from '../../../../test/mockApi/assessment/assessmentMockMutation'
import { getStudentQuerySuccess } from '../../../../test/mockApi/student/studentMockQuery'

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ id: 'studentId001' }),
}))

describe('<AssessmentCard>', () => {
  describe('@Render', () => {
    const props = {
      assessment: {
        title: 'Assessment Title',
        description: 'description',
        assessmentId: 'alphabet-upperCase',
      },
    }
    it('Should render a default state', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(
          AssessmentCard,
          [addAssessmentMutationSuccess, getStudentQuerySuccess],
          props
        )
        await delay()
      })
      wrapper.update()
      expect(wrapper.find(AssessmentCard)).toMatchSnapshot()
    })
  })
})
