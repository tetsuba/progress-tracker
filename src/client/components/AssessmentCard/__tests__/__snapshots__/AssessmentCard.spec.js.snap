// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssessmentCard> @Render Should render a default state 1`] = `
<AssessmentCard
  assessment={
    Object {
      "assessmentId": "alphabet-upperCase",
      "description": "description",
      "title": "Assessment Title",
    }
  }
>
  <Col
    key="Assessment Title"
    sm={4}
  >
    <div
      className="col-sm-4"
    >
      <div
        className="card text-body text-decoration-none"
      >
        <div
          className="card-header"
        >
          <h5>
            Assessment Title
          </h5>
        </div>
        <div
          className="card-body"
        >
          description
        </div>
        <div
          className="card-footer"
        >
          <Button
            active={false}
            className="float-right"
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="primary"
          >
            <button
              className="float-right btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      448,
                      512,
                      Array [],
                      "f067",
                      "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
                    ],
                    "iconName": "plus",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-plus fa-w-14 "
                  data-icon="plus"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
               Add Assessment
            </button>
          </Button>
        </div>
      </div>
    </div>
  </Col>
</AssessmentCard>
`;
